calyx-py/calyx/queues.py

```diff
@@ -149,37 +142,311 @@ def __len__(self) -> int:
+
+        return self.query(time, val, True, return_rank)
+
+    def peek(self, time=0, val=None, return_rank=False) -> Optional[int]:
+        """Peeks a PIEO. See query() for specifics."""
+
+        return self.query(time, val, False, return_rank)
+
+@dataclass
+class PCQ:
+    """A Programmable Calendar Queue (PCQ) data structure.
+    Supports the operations `push`, `pop`, and `peek`.
+
+    Elements are stored in buckets within which they are ordered by priority.
+    Bucket priority is determined by the 'day pointer', which points to the current
+    bucket with highest priority. 
+
+    At initialization we take in a set of `(int, int)` tuple lists `data` which stores
+    lists of values and their ranks, each representing a bucket.
+
+    We store the highest rank of any element in the queue, such that we can upper bound the set of ranks.
+
+    When asked to push:
+    - We compute the bucket to push to as the rank of the new element multiplied by bucket width,
+    wrapped around to be within the size of the queue.
+    - We then insert said new element into its assigned bucket such that the rank order is preserved.
+
+    When asked to pop:
+    - If the length of `data` is 0, we raise an error .
+    - Otherwise, we pop the lowest-rank element of the current day.
+    - If, following our pop, the bucket is empty, we rotate to the next bucket.
+
+    When asked to peek:
+    We do the same thing as `pop`, except:
+    - We peek into the PCQ instead of popping it - i.e. we don't remove any elements.
+
+    We compactly represent these similar operations through `query`, which takes in an additional
+    optional `remove` parameter (defaulted to False) to determine whether to pop or peek.
+    """
+
+    def __init__(self, max_len: int, num_buckets=200, width=10, initial_day=0):
+        self.width = width
+        self.day = initial_day
+        self.max_len = max_len
+        self.num_elements = 0
+        self.data = []
+        self.bucket_ranges = []
+        self.highest_rank = 0
+        self.insertion_count = 0
+        for i in range(num_buckets):
+            self.data.append(Pieo(16))
+            self.bucket_ranges.append((i*width, (i+1)*width))
+
+    
+    def __repr__(self):
+        final = []
+        for x in self.data:
+            if len(x.data) != 0:
+                final += x.data
+        final.sort(key = lambda k : k[2])
+        return str(final)
+    
+    def rotate(self) -> None:
+        """Rotates a PCQ and changes the 'top' parameter of the previous bucket."""
+        buckettop = self.bucket_ranges[self.day][1]
+        self.bucket_ranges[self.day] = (buckettop, buckettop + self.width)
+        self.day = (self.day + 1) % len(self.data)
+
+    def push(self, val: int, time=0, rank=0) -> None:
+        """Pushes a value with some rank/priority to a PCQ"""
+
+        if self.num_elements == self.max_len:
+            raise QueueError("Overflow")
+
+        location = (rank * self.width) % len(self.data)
```

Comment:
Is this set, or tunable?

Reply:
That is, do we understand that this is the best hashing function? Or could there be reasons to choose another?

Reply:
I'm opening an [issue](https://github.com/cucapra/packet-scheduling/issues/21) for this in the packet-scheduling repo. I think it requires a bit more thought than we can give in just this code review. For the time being I'd opt to go ahead with this, if that's ok by everyone!